File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/G-Гоблины_и_очереди/solution.py
import sys
from collections import deque


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    queue: deque[int] = deque()
    output: deque[int] = deque()
    total_size: int = 0
    for _ in range(n):
        operation, *args = sys.stdin.readline().rstrip().split()
        if operation == '-':
            total_size -= 1
            if output:
                sys.stdout.write(f'{output.popleft()}\n')
            else:
                sys.stdout.write(f'{queue.popleft()}\n')
            continue
        if operation == '+':
            queue.append(int(args[0]))
        else:
            place: int = total_size // 2
            if total_size & 1: place += 1
            place -= len(output)
            for _ in range(place):
                output.append(queue.popleft());
            queue.appendleft(int(args[0]))
        total_size += 1


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/G-Гоблины_и_очереди/solution.cpp
#include <iostream>
#include <string>
#include <deque>

using namespace std;

void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n;
    scanf("%d", &n);
    deque<int> queue;
    deque<int> output;
    int totalSize = 0;
    for (int i = 0; i < n; ++i) {
        string event;
        cin >> event;
        if (event == "-") {
            --totalSize;
            if (!output.empty()) {
                printf("%d\n", output.front());
                output.pop_front();
            } else {
                printf("%d\n", queue.front());
                queue.pop_front();
            }
            continue;
        }
        int goblinId;
        scanf("%d", &goblinId);
        if (event == "+") {
            queue.push_back(goblinId);
        } else {
            int place = totalSize / 2;
            if (totalSize % 2 != 0) ++place;
            place -= output.size();
            for (int idx = 0; idx < place; ++idx) {
                output.push_back(queue.front());
                queue.pop_front();
            }
            queue.push_front(goblinId);
        }
        ++totalSize;
    }
}


int main() {
    solution();
}